distml/operator/jax_operator.py

```diff
@@ -0,0 +1,338 @@
+import numpy as np
+import cupy as cp
+from jax import value_and_grad
+import jax.numpy as jnp
+from jax.lib import xla_client
+from jax.dlpack import from_dlpack
+from jax.tree_util import tree_flatten, tree_unflatten, tree_structure
+from jax._src.util import unzip2
+from jax.experimental.optimizers import OptimizerState
+
+from distml.operator.base_operator import TrainingOperator
+
+
+class JAXTrainingOperator(TrainingOperator):
+    def __init__(self, operator_config):
+        super(JAXTrainingOperator, self).__init__(operator_config)
+        # Should be set by users in the `register` function.
+        # model methods
+        self.opt_state = None
+        self.init_fun = None
+        self.predict_fun = None
+        # optimizer methods
+        self.opt_init = None
+        self.opt_update = None
+        self.get_params = None
+
+        self.criterion = None
+
+        # Data loaders for training and validation, registered by users.
+        self._train_loader = None
+        self._validation_loader = None
+
+        self.setup(operator_config)
+
+        if hasattr(operator_config, "jit_mode"):
+            if operator_config["jit_mode"]:
+                raise NotImplementedError("Not support jit in jax operator.")
+
+        self.train_step_num = 0
+
+    def setup(self, *args, **kwargs):
+        """Function that needs to be override by users.
+
+        Example:
+            # some code is the same for all users,
+            # maybe we can put it in register.
+            rng_key = random.PRNGKey(0)
+            input_shape = (28, 28, 1, batch_size)
+            lr=0.01
+            init_fun, predict_fun = ResNet18(num_classes)
+            _, init_params = init_fun(rng_key, input_shape)
+
+            opt_init, opt_update, get_params = optimizers.adam(lr)
+            opt_state = opt_init(init_params)
+
+            criterion = lambda logits, targets:-jnp.sum(logits * targets)
+
+            self.register(model=(opt_state, init_fun, predict_fun),
+                          optimizer=(opt_init, opt_update, get_params),
+                          criterion=criterion)
+        """
+        raise NotImplementedError("Please override this function to register "
+                                  "your model, optimizer, and criterion.")
+
+    def register(self,
+                 *,
+                 model,
+                 optimizer,
+                 criterion,
+                 jit_mode=False):
+        """Register a few critical information about the model to operator."""
+        self.criterion = criterion
```

Comment:
@Ezra-H : please add a sanity check here to make sure `model`, `optimizer` and `critierion` are valida JAX types. See TorchTrainingOperator for examples.